src/client/gfs_client.cc

```diff
@@ -1,32 +1,80 @@
 #include "src/client/gfs_client.h"
+#include "src/client/client_impl.h"
+#include "src/common/utils.h"
 
 namespace gfs {
 namespace client {
 
-google::protobuf::util::Status open(const char* pathname, unsigned int flags) {
-  // Step 1. Check and validate the flags, e.g. it doesn't make sense to 
-  // open with both read and write flags
+// A singleton object that serves as the interface for the impl code to access
+// and issue calls to internal units such as cache manager, service clients to
+// master and chunk servers
+static ClientImpl* client_impl_ = nullptr;
 
-  // Creation mode
-  if (flags | OpenFlag::Create) {
+// Validate a given open flag, currently we only support four modes, namely
+// read, write, create and create | write
+bool ValidateOpenFlag(unsigned int flags) {
+  return flags == OpenFlag::Read || flags == OpenFlag::Write ||
+         flags == OpenFlag::Create || 
+         flags == (OpenFlag::Write | OpenFlag::Create);
```

Comment:
nit;  currently, write has a flag value of 3 and create has a flag value of 4, so it's not safe if in future we have a field in protobuf definition with a tag value of 7 (= 3 | 4), (even tho it won't happen for now in our case); it's better to "reserve" the field value "7" in proto definition (see protobuf syntax doc)

Reply:
I guess I confused you with OpenFlag type and the enum type defined in proto. The OpenFlag enum (see gfs_client.h) defines the flags as 0x1, 0x2, and 0x4 which are bit masks and are different from the types in proto def. I presume the proto types are transparent from the users.